ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -0,0 +1,599 @@
+        let divisor = UInt::new_unchecked(|| "divisor", cb)?; // 32-bit value from rs2
+        let quotient = UInt::new(|| "quotient", cb)?;
+        let remainder = UInt::new(|| "remainder", cb)?;
+
+        // `rem_e` and `div_e` are expressions representing the remainder and
+        // divisor from the signed or unsigned division operation, with signs
+        // renormalized to be nonnegative, so that correct values must satisfy
+        // either `0 <= rem_e < div_e` or `div_e == 0`.  The `rem_e` value
+        // should be constrained to be nonnegative before being returned from
+        // this block, while the checks `rem_e < div_e` or `div_e == 0` are
+        // done later.
+        let (internal_config, rem_e, div_e) = match I::INST_KIND {
+            InsnKind::DIVU | InsnKind::REMU => {
+                cb.require_equal(
+                    || "unsigned_division_relation",
+                    dividend.value(),
+                    divisor.value() * quotient.value() + remainder.value(),
+                )?;
+
+                (InternalDivRem::Unsigned, remainder.value(), divisor.value())
+            }
+
+            InsnKind::DIV | InsnKind::REM => {
+                let dividend_signed: Signed<E> =
+                    Signed::construct_circuit(cb, || "dividend_signed", &dividend)?;
+                let divisor_signed: Signed<E> =
+                    Signed::construct_circuit(cb, || "divisor_signed", &divisor)?;
+                let quotient_signed: Signed<E> =
+                    Signed::construct_circuit(cb, || "quotient_signed", &quotient)?;
+                let remainder_signed: Signed<E> =
+                    Signed::construct_circuit(cb, || "remainder_signed", &quotient)?;
+
+                // The quotient and remainder can be interpreted as non-positive
+                // values when exactly one of dividend and divisor is negative
+                let neg_div_expr: Expression<E> = {
+                    let a_neg = dividend_signed.is_negative.expr();
+                    let b_neg = divisor_signed.is_negative.expr();
+                    // a_neg * (1 - b_neg) + (1 - a_neg) * b_neg
+                    (a_neg.clone() + b_neg.clone()) - (Expression::<E>::from(2) * a_neg * b_neg)
+                };
+                let negative_division = cb.flatten_expr(|| "neg_division", neg_div_expr)?;
+
+                // Check for signed division overflow: i32::MIN / -1
+                let is_dividend_max_negative = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_dividend_max_negative",
+                    dividend.value(),
+                    (1u64 << (BIT_WIDTH - 1)).into(),
+                )?;
+                let is_divisor_minus_one = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_divisor_minus_one",
+                    divisor.value(),
+                    ((1u64 << BIT_WIDTH) - 1).into(),
```

Comment:
This curcuit is limited to 32-bits registers and Goldiliock field so I think we can just use `u32::MAX` here.

Reply:
Yep, that's fair, the limitation is pretty hard-coded throughout, and this is definitely more readable.  Will make the change.